_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,429 @@
+
+If a public method named `setUp()` is present it will be executed before each test is run.
+
+Three types of test are recognised by `hevm`:
+
+| type           | prefix  | arguments | semantics                                                   |
+|----------------|---------|-----------|-------------------------------------------------------------|
+| concrete       | `test`  | no        | single execution with concrete values                       |
+| property based | `test`  | yes       | multiple executions with randomly generated concrete values |
+| symbolic       | `prove` | optional  | exhaustive exploration of all possible exeuction paths      |
+
+To illustrate the differences between the test types, consider the following example:
+
+```solidity
+contract Test is DSTest {
+    function test_associativity() public {
+        assertEq(
+            uint((1 + 2) + 3),
+            uint(1 + (2 + 3))
+        );
+    }
+
+    function test_associativity_fuzz(uint x, uint y, uint z) public {
+        assertEq(
+            (x + y) + z,
+            x + (y + z)
+        );
+    }
+
+    function prove_associativity(uint x, uint y, uint z) public {
+        assertEq(
+          (x + y) + z,
+          x + (y + z)
+        );
+    }
+}
+```
+
+- `test_associativity` will run a single test case with the concrete values of `1`, `2`, and `3`.
+- `test_associativity_fuzz` will be executed many times (100 by default), with randomly generated values for `x`, `y`, and `z`.
+- `prove_associativity` will be symbolically executed, with `x`, `y`, and `z` represented as symbolic variables.
+
+Finally, it is possible to manipulate the execution environment (e.g. timestamp) from within
+`ds-test` using by using hevm "[cheat codes](https://github.com/dapphub/dapptools/tree/master/src/hevm#cheat-codes)"
+
+## Finding Counterexamples
+
+Let's look at a more complex example! Consider the following token contract:
+
+```solidity
+pragma solidity ^0.6.12;
+
+contract SafeMath {
+    function add(uint x, uint y) internal pure returns (uint z) {
+        require((z = x + y) >= x, "overflow");
+    }
+    function sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x, "underflow");
+    }
+}
+
+contract Token is SafeMath {
+    uint256 public totalSupply;
+    mapping (address => uint) public balanceOf;
+
+    constructor(uint _totalSupply) public {
+        totalSupply           = _totalSupply;
+        balanceOf[msg.sender] = _totalSupply;
+    }
+
+    function transfer(address dst, uint amt) public {
+        balanceOf[msg.sender] = sub(balanceOf[msg.sender], amt);
+        balanceOf[dst]        = add(balanceOf[dst], amt);
+    }
+}
+```
+
+We can write a test that asserts our expected behaviour for the transfer function as follows:
+
+```solidity
+contract TestToken is DSTest, SafeMath {
+    function prove_transfer(address dst, uint amt) public {
+        Token token = new Token(uint(-1));
```

Comment:
```suggestion
        Token token = new Token(type(uint).max);
```

Reply:
I don't have  a super strong opinion about it, but I actually find the `uint(-1)` form slightly easier to read, and iirc it will generate slightly more efficient code if the optimizer isn't enabled.

Reply:
Easier than explicitly saying `max of type`? Why would it generate more efficient code? Both translate to the same number. Besides, `uint(-1)` is disallowed from 0.8

Reply:
Plus `uint(-1)` relies on underflow (conceptually) which is unrelated to simply using a constant, which IMO makes it less appropriate

Reply:
> Why would it generate more efficient code?

I was actaully wrong here. I was misremebering a twitter convo where it emerged that solidity was inserting some unescessary instructions when accessing `contract.functionName.selector` without the optimizer enabled.

Anyway, I'm happy to have `type(uint).max`. I updated the post to use that instead.